lib/phoenix/channel/server.ex

```diff
@@ -314,11 +339,33 @@ defmodule Phoenix.Channel.Server do
         %{topic: topic} = socket
       ) do
     start = System.monotonic_time()
-    result = socket.channel.handle_in(event, payload, put_in(socket.ref, ref))
-    duration = System.monotonic_time() - start
-    metadata = %{ref: ref, event: event, params: payload, socket: socket}
-    :telemetry.execute([:phoenix, :channel_handled_in], %{duration: duration}, metadata)
-    handle_in(result)
+    metadata = %{event: event, params: payload, socket: socket, ref: ref}
+    measurements = %{system_time: System.system_time()}
+    :telemetry.execute([:phoenix, :channel, :handle_in, :start], measurements, metadata)
+
+    try do
+      result = socket.channel.handle_in(event, payload, put_in(socket.ref, ref))
+      state = handle_in(result)
+
+      measurements = %{duration: System.monotonic_time() - start}
+
+      :telemetry.execute(
+        [:phoenix, :channel, :handle_in, :stop],
+        measurements,
+        Map.merge(metadata, %{socket: get_socket(state), result: result})
```

Comment:
Having the full return value of `handle_in/3` in the metadata is very helpful. I'm not sure about naming though, because the `[.., :join, :stop]` event also has the `:result` key, but it's either `:ok` or `:error` (to be consistent with the existing `:channel_joined` event).